### DISCUSSION TITLE
Mirth Converts MSH|^~\&| to MSH|^~\&amp:|

### DISCUSSION THREAD

**Discussion by LanciaH** [nextgenhealthcare/connect#5146] [Q&A]
Hi All, First post, apologies if this has been answered before.  I tried searching and found posts kinda related but not exactly my problem.

This is our first time using Mirth, We are using Mirth to convert a CSV file to HL7 file and then send the HL7 file through a web service in a SOAP wrapper.

The HL7 file get created fine and has been validated.  Looks like when MIRTH converts the HL7 file to XML, MIRTH converts & to &amp: from MSH|^~\**&**| to <MSH.2>^~\**&amp:**</MSH.2>  causing problems in the webservice file transfer.  At least that is what the vendor tells me.

Is there a way to prevent this conversion of & to &amp:?

Thanks,
Ralph

![image](https://user-images.githubusercontent.com/104143486/164456699-0ab2a5c8-8985-4c18-aac7-c50f7ad72e0a.png)

**Comment by jonbartels** [comment]
> Hi All, First post, apologies if this has been answered before. I tried searching and found posts kinda related but not exactly my problem.

Great first post. Welcome to the community. Please share what you searched for and where you looked. Github discussions are good. https://forums.mirthproject.io/ has decades of old threads.

Please also make sure your screenshots do not contain real patient information.

If you can, please also edit your post to include:
* XML snippets of your inbound and outbound messages. Use the code formatting tags please
* Show your "data types" settings from the MC channel

This is definitely an XML escaping issue but I can't quite see where the problem is. The inbound `RequestHL7Message` has an ampersand in it but it is XML, if you escape the inbound message to have `&amp;` does your HL7 render correctly?

**Comment by LanciaH** [comment]
Hi Jo,.
In Google I searched for the following MIRTH &amp: and received the following results which led me to this forum.
I did not setup the channel but I'll take a look.  Not sure how to do this part, "if you escape the inbound message to have &amp;"

Channel Data Types?
![image](https://user-images.githubusercontent.com/104143486/164463134-b63a9ca7-b7c4-4b73-8fa0-25960b2df961.png)

Raw Message
MSH|^~\&|NJDOCCentricity|9999|NJIIS|NJDOH|2022042014720||VXU^V04^VXU_V04|NJDOCCentricity64412022042014720|T|2.5.1|||NE|AL|||||Z22^CDCPHINVS|

Transformed & Encoded
<HL7Message>
    <MSH>
        <MSH.1>|</MSH.1>
        <MSH.2>^~\&amp;</MSH.2>
        <MSH.3>
            <MSH.3.1>NJDOCCentricity</MSH.3.1>
        </MSH.3>
        <MSH.4>
            <MSH.4.1>9999</MSH.4.1>
        </MSH.4>
        <MSH.5>
            <MSH.5.1>NJIIS</MSH.5.1>
        </MSH.5>
        <MSH.6>
            <MSH.6.1>NJDOH</MSH.6.1>
        </MSH.6>
        <MSH.7>
            <MSH.7.1>2022042014720</MSH.7.1>

Part of the Sent message, I can't show all of it due to medical info
WSDL URL: https://njiis-uat.nj.gov:443/ims/service?wsdl
SERVICE: {http://njiis.njdoh.gov/ims/service}NJIISIMSService
PORT / ENDPOINT: {http://njiis.njdoh.gov/ims/service}NJIISIMSServicePort
LOCATION URI: https://njiis-uat.nj.gov:443/ims/service
SOAP ACTION: http://njiis.nj.gov/ims/service/submitPatientImmInfo

[ATTACHMENTS]

[CONTENT]
<soap:Envelope xmlns:soap="http://www.w3.org/2003/05/soap-envelope" xmlns:req="http://njiis.njdoh.gov/ims/service/schema/hl7/put/request">
   <soap:Header/>
   <soap:Body>
      <req:NJIISIMSPutRequest>
         <req:RequestHeader>
            <req:FacilityID>9999999</req:FacilityID>
            <req:FacilityKey>999999</req:FacilityKey>
         </req:RequestHeader>
         <req:PutRequest>
            
            <req:NJIISProviderID>9999</req:NJIISProviderID>
            <req:HL7MessageVersion>2.5.1</req:HL7MessageVersion>
            <req:RequestHL7Message>MSH|^~\&|NJDOCCentricity|9999|NJIIS|NJDOH|2022042014720||VXU^V04^VXU_V04|NJDOCCentricity64412022042014720|T|2.5.1|||NE|AL|||||Z22^CDCPHINVS|

Response Error 
ERROR: Error creating web service dispatch [SOAPExceptionImpl: Error during saving a multipart message]

![image](https://user-images.githubusercontent.com/104143486/164462446-233cf827-691c-42b7-a053-647b1dde3f93.png)

**Reply by jonbartels** [reply]
Excuse me I read the use case backwards. CSV input, to HL7, sending in a SOAP payload.

Can you show the JS or mapper step where you populate `RequestHL7Message` - It looks like the HL7 message as a string is being inserted directly to the outbound SOAP XML without being escaped first. The outbound message is then failing because of the escaping. I think you might just have to escape the HL7 message string before inserting it to the SOAP payload.

The inbound message having `<MSH.2>^~&amp;</MSH.2>` is correct. HL7 has a string `MSH|^~&|NJDOCCentricity|9999|NJIIS|N...` then the internal HL7 message structure is XML. The raw `&` gets escaped to `&amp;` in XML. Then it gets turned back into `&` when the XML is printed with `toString`. Finally when `MSH|^~&|NJDOCCentricity|9999|NJIIS|N` is inserted to the SOAP XML the SOAP XML sender complains because `&` isn't escaped. 

I hope I have understood the problem and explained the chain of events properly.

**Reply by LanciaH** [reply]
Hi Jon,
Yes II think you understand the problem completely, I'm a noob to all this.  I feel we are close but just stuck. Anyway, is below what you are looking for?

tmp['MSH']['MSH.1']=msg['MSH'][0]['MSH.1'].toString()
//tmp['MSH']['MSH.2']=msg['MSH'][0]['MSH.2'].toString()
tmp['MSH']['MSH.2']="^~\\&amp;"
tmp['MSH']['MSH.3']['MSH.3.1']=msg['MSH'][0]['MSH.3']['MSH.3.1'].toString()
tmp['MSH']['MSH.4']['MSH.4.1']=msg['MSH'][0]['MSH.4']['MSH.4.1'].toString()
tmp['MSH']['MSH.5']['MSH.5.1']=msg['MSH'][0]['MSH.5']['MSH.5.1'].toString()
tmp['MSH']['MSH.6']['MSH.6.1']=msg['MSH'][0]['MSH.6']['MSH.6.1'].toString()
tmp['MSH']['MSH.7']['MSH.7.1']=msg['MSH'][0]['MSH.7']['MSH.7.1'].toString()
tmp['MSH']['MSH.9']['MSH.9.1']=msg['MSH'][0]['MSH.9']['MSH.9.1'].toString()
tmp['MSH']['MSH.9']['MSH.9.2']=msg['MSH'][0]['MSH.9']['MSH.9.2'].toString()
tmp['MSH']['MSH.9']['MSH.9.3']=msg['MSH'][0]['MSH.9']['MSH.9.3'].toString()
tmp['MSH']['MSH.10']['MSH.10.1']=msg['MSH'][0]['MSH.10']['MSH.10.1'].toString()
tmp['MSH']['MSH.11']['MSH.11.1']=msg['MSH'][0]['MSH.11']['MSH.11.1'].toString()
tmp['MSH']['MSH.12']['MSH.12.1']=msg['MSH'][0]['MSH.12']['MSH.12.1'].toString()
tmp['MSH']['MSH.15']['MSH.15.1']=msg['MSH'][0]['MSH.15']['MSH.15.1'].toString()
tmp['MSH']['MSH.16']['MSH.16.1']=msg['MSH'][0]['MSH.16']['MSH.16.1'].toString()
tmp['MSH']['MSH.21']['MSH.21.1']=msg['MSH'][0]['MSH.21']['MSH.21.1'].toString()
tmp['MSH']['MSH.21']['MSH.21.2']=msg['MSH'][0]['MSH.21']['MSH.21.2'].toString()
tmp['PID']['PID.1']['PID.1.1']=msg['PID'][0]['PID.1']['PID.1.1'].toString()
tmp['PID']['PID.3']['PID.3.1']=msg['PID'][0]['PID.3']['PID.3.1'].toString()
tmp['PID']['PID.3']['PID.3.2']=msg['PID'][0]['PID.3']['PID.3.2'].toString()
tmp['PID']['PID.3']['PID.3.3']=msg['PID'][0]['PID.3']['PID.3.3'].toString()
tmp['PID']['PID.3']['PID.3.4']=msg['PID'][0]['PID.3']['PID.3.4'].toString()
tmp['PID']['PID.3']['PID.3.5']=msg['PID'][0]['PID.3']['PID.3.5'].toString()
tmp['PID']['PID.5']['PID.5.1']=msg['PID'][0]['PID.5']['PID.5.1'].toString()
tmp['PID']['PID.5']['PID.5.2']=msg['PID'][0]['PID.5']['PID.5.2'].toString()
tmp['PID']['PID.5']['PID.5.3']=msg['PID'][0]['PID.5']['PID.5.3'].toString()
tmp['PID']['PID.5']['PID.5.4']=msg['PID'][0]['PID.5']['PID.5.4'].toString()
tmp['PID']['PID.5']['PID.5.5']=msg['PID'][0]['PID.5']['PID.5.5'].toString()
tmp['PID']['PID.5']['PID.5.6']=msg['PID'][0]['PID.5']['PID.5.6'].toString()
tmp['PID']['PID.5']['PID.5.7']=msg['PID'][0]['PID.5']['PID.5.7'].toString()
tmp['PID']['PID.7']['PID.7.1']=msg['PID'][0]['PID.7']['PID.7.1'].toString()
tmp['PID']['PID.8']['PID.8.1']=msg['PID'][0]['PID.8']['PID.8.1'].toString()
tmp['PID']['PID.11']['PID.11.1']=msg['PID'][0]['PID.11']['PID.11.1'].toString()
tmp['PID']['PID.11']['PID.11.2']=msg['PID'][0]['PID.11']['PID.11.2'].toString()
tmp['PID']['PID.11']['PID.11.3']=msg['PID'][0]['PID.11']['PID.11.3'].toString()
tmp['PID']['PID.11']['PID.11.4']=msg['PID'][0]['PID.11']['PID.11.4'].toString()
tmp['PID']['PID.11']['PID.11.5']=msg['PID'][0]['PID.11']['PID.11.5'].toString()
tmp['PID']['PID.11']['PID.11.6']=msg['PID'][0]['PID.11']['PID.11.6'].toString()
tmp['PID']['PID.11']['PID.11.7']=msg['PID'][0]['PID.11']['PID.11.7'].toString()
tmp['PD1']['PD1.12']['PD1.12.1']=msg['PD1'][0]['PD1.12']['PD1.12.1'].toString()
tmp['PD1']['PD1.13']['PD1.13.1']=msg['PD1'][0]['PD1.13']['PD1.13.1'].toString()
tmp['NK1']['NK1.1']['NK1.1.1']=msg['NK1'][0]['NK1.1']['NK1.1.1'].toString()
tmp['NK1']['NK1.2']['NK1.2.1']=msg['NK1'][0]['NK1.2']['NK1.2.1'].toString()
tmp['NK1']['NK1.2']['NK1.2.2']=msg['NK1'][0]['NK1.2']['NK1.2.2'].toString()
tmp['NK1']['NK1.2']['NK1.2.3']=msg['NK1'][0]['NK1.2']['NK1.2.3'].toString()
tmp['NK1']['NK1.2']['NK1.2.4']=msg['NK1'][0]['NK1.2']['NK1.2.4'].toString()
tmp['NK1']['NK1.2']['NK1.2.5']=msg['NK1'][0]['NK1.2']['NK1.2.5'].toString()
tmp['NK1']['NK1.3']['NK1.3.1']=msg['NK1'][0]['NK1.3']['NK1.3.1'].toString()
tmp['NK1']['NK1.3']['NK1.3.2']=msg['NK1'][0]['NK1.3']['NK1.3.2'].toString()
tmp['NK1']['NK1.3']['NK1.3.3']=msg['NK1'][0]['NK1.3']['NK1.3.3'].toString()
tmp['NK1']['NK1.4']['NK1.4.1']=msg['NK1'][0]['NK1.4']['NK1.4.1'].toString()
tmp['NK1']['NK1.4']['NK1.4.2']=msg['NK1'][0]['NK1.4']['NK1.4.2'].toString()
tmp['NK1']['NK1.4']['NK1.4.3']=msg['NK1'][0]['NK1.4']['NK1.4.3'].toString()
tmp['NK1']['NK1.4']['NK1.4.4']=msg['NK1'][0]['NK1.4']['NK1.4.4'].toString()
tmp['NK1']['NK1.4']['NK1.4.5']=msg['NK1'][0]['NK1.4']['NK1.4.5'].toString()
tmp['NK1']['NK1.4']['NK1.4.6']=msg['NK1'][0]['NK1.4']['NK1.4.6'].toString()
tmp['NK1']['NK1.4']['NK1.4.7']=msg['NK1'][0]['NK1.4']['NK1.4.7'].toString()
tmp['ORC'][0]['ORC.1']['ORC.1.1']=msg['ORC'][0]['ORC.1']['ORC.1.1'].toString()
tmp['ORC'][0]['ORC.2']['ORC.2.1']=msg['ORC'][0]['ORC.2']['ORC.2.1'].toString()
tmp['ORC'][0]['ORC.3']['ORC.3.1']=msg['ORC'][0]['ORC.3']['ORC.3.1'].toString()
tmp['ORC'][0]['ORC.3']['ORC.3.2']=msg['ORC'][0]['ORC.3']['ORC.3.2'].toString()
tmp['ORC'][0]['ORC.3']['ORC.3.3']=msg['ORC'][0]['ORC.3']['ORC.3.3'].toString()
tmp['ORC'][0]['ORC.3']['ORC.3.4']=msg['ORC'][0]['ORC.3']['ORC.3.4'].toString()
tmp['RXA'][0]['RXA.1']['RXA.1.1']=msg['RXA'][0]['RXA.1']['RXA.1.1'].toString()
tmp['RXA'][0]['RXA.2']['RXA.2.1']=msg['RXA'][0]['RXA.2']['RXA.2.1'].toString()
tmp['RXA'][0]['RXA.3']['RXA.3.1']=msg['RXA'][0]['RXA.3']['RXA.3.1'].toString()
tmp['RXA'][0]['RXA.5']['RXA.5.1']=msg['RXA'][0]['RXA.5']['RXA.5.1'].toString()
tmp['RXA'][0]['RXA.5']['RXA.5.2']=msg['RXA'][0]['RXA.5']['RXA.5.2'].toString()
tmp['RXA'][0]['RXA.5']['RXA.5.3']=msg['RXA'][0]['RXA.5']['RXA.5.3'].toString()
tmp['RXA'][0]['RXA.6']['RXA.6.1']=msg['RXA'][0]['RXA.6']['RXA.6.1'].toString()
tmp['RXA'][0]['RXA.7']['RXA.7.1']=msg['RXA'][0]['RXA.7']['RXA.7.1'].toString()
tmp['RXA'][0]['RXA.7']['RXA.7.3']=msg['RXA'][0]['RXA.7']['RXA.7.3'].toString()
tmp['RXA'][0]['RXA.9']['RXA.9.1']=msg['RXA'][0]['RXA.9']['RXA.9.1'].toString()
tmp['RXA'][0]['RXA.9']['RXA.9.2']=msg['RXA'][0]['RXA.9']['RXA.9.2'].toString()
tmp['RXA'][0]['RXA.9']['RXA.9.3']=msg['RXA'][0]['RXA.9']['RXA.9.3'].toString()
tmp['RXA'][0]['RXA.10']['RXA.10.1']=msg['RXA'][0]['RXA.10']['RXA.10.1'].toString()
tmp['RXA'][0]['RXA.11']['RXA.11.4']=msg['RXA'][0]['RXA.11']['RXA.11.4'].toString()
tmp['RXA'][0]['RXA.15']['RXA.15.1']=msg['RXA'][0]['RXA.15']['RXA.15.1'].toString()
tmp['RXA'][0]['RXA.17']['RXA.17.1']=msg['RXA'][0]['RXA.17']['RXA.17.1'].toString()
tmp['RXA'][0]['RXA.17']['RXA.17.2']=msg['RXA'][0]['RXA.17']['RXA.17.2'].toString()
tmp['RXA'][0]['RXA.17']['RXA.17.3']=msg['RXA'][0]['RXA.17']['RXA.17.3'].toString()
tmp['RXA'][0]['RXA.20']['RXA.20.1']=msg['RXA'][0]['RXA.20']['RXA.20.1'].toString()
tmp['RXA'][0]['RXA.21']['RXA.21.1']=msg['RXA'][0]['RXA.21']['RXA.21.1'].toString()
tmp['RXR'][0]['RXR.1']['RXR.1.1']=msg['RXR'][0]['RXR.1']['RXR.1.1'].toString()
tmp['RXR'][0]['RXR.1']['RXR.1.2']=msg['RXR'][0]['RXR.1']['RXR.1.2'].toString()
tmp['RXR'][0]['RXR.1']['RXR.1.3']=msg['RXR'][0]['RXR.1']['RXR.1.3'].toString()
tmp['RXR'][0]['RXR.1']['RXR.1.4']=msg['RXR'][0]['RXR.1']['RXR.1.4'].toString()
tmp['RXR'][0]['RXR.2']['RXR.2.1']=msg['RXR'][0]['RXR.2']['RXR.2.1'].toString()
tmp['RXR'][0]['RXR.2']['RXR.2.2']=msg['RXR'][0]['RXR.2']['RXR.2.2'].toString()
tmp['RXR'][0]['RXR.2']['RXR.2.3']=msg['RXR'][0]['RXR.2']['RXR.2.3'].toString()
tmp['RXR'][0]['RXR.2']['RXR.2.4']=msg['RXR'][0]['RXR.2']['RXR.2.4'].toString()
tmp['RXR'][0]['RXR.3']['RXR.3.1']=msg['RXR'][0]['RXR.3']['RXR.3.1'].toString()
tmp['OBX'][0]['OBX.1']['OBX.1.1']=msg['OBX'][0]['OBX.1']['OBX.1.1'].toString()
tmp['OBX'][0]['OBX.2']['OBX.2.1']=msg['OBX'][0]['OBX.2']['OBX.2.1'].toString()
tmp['OBX'][0]['OBX.3']['OBX.3.1']=msg['OBX'][0]['OBX.3']['OBX.3.1'].toString()
tmp['OBX'][0]['OBX.3']['OBX.3.2']=msg['OBX'][0]['OBX.3']['OBX.3.2'].toString()
tmp['OBX'][0]['OBX.3']['OBX.3.3']=msg['OBX'][0]['OBX.3']['OBX.3.3'].toString()
tmp['OBX'][0]['OBX.4']['OBX.4.1']=msg['OBX'][0]['OBX.4']['OBX.4.1'].toString()
tmp['OBX'][0]['OBX.5']['OBX.5.1']=msg['OBX'][0]['OBX.5']['OBX.5.1'].toString()
tmp['OBX'][0]['OBX.5']['OBX.5.2']=msg['OBX'][0]['OBX.5']['OBX.5.2'].toString()
tmp['OBX'][0]['OBX.5']['OBX.5.3']=msg['OBX'][0]['OBX.5']['OBX.5.3'].toString()
tmp['OBX'][0]['OBX.11']['OBX.11.1']=msg['OBX'][0]['OBX.11']['OBX.11.1'].toString()
tmp['ORC'][1]['ORC.1']['ORC.1.1']=msg['ORC'][1]['ORC.1']['ORC.1.1'].toString()
tmp['ORC'][1]['ORC.2']['ORC.2.1']=msg['ORC'][1]['ORC.2']['ORC.2.1'].toString()
tmp['ORC'][1]['ORC.3']['ORC.3.1']=msg['ORC'][1]['ORC.3']['ORC.3.1'].toString()
tmp['ORC'][1]['ORC.3']['ORC.3.2']=msg['ORC'][1]['ORC.3']['ORC.3.2'].toString()
tmp['ORC'][1]['ORC.3']['ORC.3.3']=msg['ORC'][1]['ORC.3']['ORC.3.3'].toString()
tmp['ORC'][1]['ORC.3']['ORC.3.4']=msg['ORC'][1]['ORC.3']['ORC.3.4'].toString()
tmp['ORC'][1]['ORC.4']['ORC.4.1']=msg['ORC'][1]['ORC.4']['ORC.4.1'].toString()
tmp['RXA'][1]['RXA.1']['RXA.1.1']=msg['RXA'][1]['RXA.1']['RXA.1.1'].toString()
tmp['RXA'][1]['RXA.2']['RXA.2.1']=msg['RXA'][1]['RXA.2']['RXA.2.1'].toString()
tmp['RXA'][1]['RXA.3']['RXA.3.1']=msg['RXA'][1]['RXA.3']['RXA.3.1'].toString()
tmp['RXA'][1]['RXA.5']['RXA.5.1']=msg['RXA'][1]['RXA.5']['RXA.5.1'].toString()
tmp['RXA'][1]['RXA.5']['RXA.5.2']=msg['RXA'][1]['RXA.5']['RXA.5.2'].toString()
tmp['RXA'][1]['RXA.5']['RXA.5.3']=msg['RXA'][1]['RXA.5']['RXA.5.3'].toString()
tmp['RXA'][1]['RXA.6']['RXA.6.1']=msg['RXA'][1]['RXA.6']['RXA.6.1'].toString()
tmp['RXA'][1]['RXA.7']['RXA.7.1']=msg['RXA'][1]['RXA.7']['RXA.7.1'].toString()
tmp['RXA'][1]['RXA.7']['RXA.7.3']=msg['RXA'][1]['RXA.7']['RXA.7.3'].toString()
tmp['RXA'][1]['RXA.9']['RXA.9.1']=msg['RXA'][1]['RXA.9']['RXA.9.1'].toString()
tmp['RXA'][1]['RXA.9']['RXA.9.2']=msg['RXA'][1]['RXA.9']['RXA.9.2'].toString()
tmp['RXA'][1]['RXA.9']['RXA.9.3']=msg['RXA'][1]['RXA.9']['RXA.9.3'].toString()
tmp['RXA'][1]['RXA.10']['RXA.10.1']=msg['RXA'][1]['RXA.10']['RXA.10.1'].toString()
tmp['RXA'][1]['RXA.11']['RXA.11.4']=msg['RXA'][1]['RXA.11']['RXA.11.4'].toString()
tmp['RXA'][1]['RXA.15']['RXA.15.1']=msg['RXA'][1]['RXA.15']['RXA.15.1'].toString()
tmp['RXA'][1]['RXA.17']['RXA.17.1']=msg['RXA'][1]['RXA.17']['RXA.17.1'].toString()
tmp['RXA'][1]['RXA.17']['RXA.17.2']=msg['RXA'][1]['RXA.17']['RXA.17.2'].toString()
tmp['RXA'][1]['RXA.17']['RXA.17.3']=msg['RXA'][1]['RXA.17']['RXA.17.3'].toString()
tmp['RXA'][1]['RXA.20']['RXA.20.1']=msg['RXA'][1]['RXA.20']['RXA.20.1'].toString()
tmp['RXA'][1]['RXA.21']['RXA.21.1']=msg['RXA'][1]['RXA.21']['RXA.21.1'].toString()
tmp['RXR'][1]['RXR.1']['RXR.1.1']=msg['RXR'][1]['RXR.1']['RXR.1.1'].toString()
tmp['RXR'][1]['RXR.1']['RXR.1.2']=msg['RXR'][1]['RXR.1']['RXR.1.2'].toString()
tmp['RXR'][1]['RXR.1']['RXR.1.3']=msg['RXR'][1]['RXR.1']['RXR.1.3'].toString()
tmp['RXR'][1]['RXR.1']['RXR.1.4']=msg['RXR'][1]['RXR.1']['RXR.1.4'].toString()
tmp['RXR'][1]['RXR.2']['RXR.2.1']=msg['RXR'][1]['RXR.2']['RXR.2.1'].toString()
tmp['RXR'][1]['RXR.2']['RXR.2.2']=msg['RXR'][1]['RXR.2']['RXR.2.2'].toString()
tmp['RXR'][1]['RXR.2']['RXR.2.3']=msg['RXR'][1]['RXR.2']['RXR.2.3'].toString()
tmp['RXR'][1]['RXR.2']['RXR.2.4']=msg['RXR'][1]['RXR.2']['RXR.2.4'].toString()
tmp['OBX'][1]['OBX.1']['OBX.1.1']=msg['OBX'][1]['OBX.1']['OBX.1.1'].toString()
tmp['OBX'][1]['OBX.2']['OBX.2.1']=msg['OBX'][1]['OBX.2']['OBX.2.1'].toString()
tmp['OBX'][1]['OBX.3']['OBX.3.1']=msg['OBX'][1]['OBX.3']['OBX.3.1'].toString()
tmp['OBX'][1]['OBX.3']['OBX.3.2']=msg['OBX'][1]['OBX.3']['OBX.3.2'].toString()
tmp['OBX'][1]['OBX.3']['OBX.3.3']=msg['OBX'][1]['OBX.3']['OBX.3.3'].toString()
tmp['OBX'][1]['OBX.4']['OBX.4.1']=msg['OBX'][1]['OBX.4']['OBX.4.1'].toString()
tmp['OBX'][1]['OBX.5']['OBX.5.1']=msg['OBX'][1]['OBX.5']['OBX.5.1'].toString()
tmp['OBX'][1]['OBX.5']['OBX.5.2']=msg['OBX'][1]['OBX.5']['OBX.5.2'].toString()
tmp['OBX'][1]['OBX.5']['OBX.5.3']=msg['OBX'][1]['OBX.5']['OBX.5.3'].toString()
tmp['OBX'][1]['OBX.11']['OBX.11.1']=msg['OBX'][1]['OBX.11']['OBX.11.1'].toString()
$c("sendMe",tmp);

![image](https://user-images.githubusercontent.com/104143486/164525187-3758a35a-0a26-4232-8fd0-aa3e9d340179.png)

**Reply by jonbartels** [reply]
How are you building the SOAP payload ?

**Reply by tonygermano** [reply]
This does not appear to be the problem in your original screenshot, but from this code, `tmp['MSH']['MSH.2']="^~\&"` should be `tmp['MSH']['MSH.2']="^~\\&"`. In javascript strings you need to escape the backslash.

There's also a good chance that most of that code is not needed, but I'm not sure what your goal is.

**Reply by LanciaH** [reply]
Jon, The SOAP envelop was built automatically in MIRTH connecting to their web service. 

Tony, We added the extra slash, thanks to both of you for the suggestion.  Transformed looks good now.
    <MSH>
        <MSH.1>|</MSH.1>
        <MSH.2>^~\&amp;amp;</MSH.2>
        <MSH.3>
            <MSH.3.1>NJDOCCentricity</MSH.3.1>


We are still getting the following server error after we send the message ERROR: Error creating web service dispatch [SOAPExceptionImpl: Error during saving a multipart message]  The vendor mentioned line breaks in the HL7 file but it looks okay and passed their HL7 validation.

The project is to extract medical data from the EHR into a CSV file, Convert that to HL7 and send the HL7 data in a SOAP envelope to their web service.

Again thanks for the help

**Reply by tonygermano** [reply]
The `\` is still missing in that XML snippet, and your `&` is double escaped. For the double escape to be correct, I would expect to see all of the `<` escaped as well.

**Reply by LanciaH** [reply]
Hi Tony, Not sure what you mean by , "I would expect to see all of the < escaped as well."

Thanks for trying to help.  I'm about to give up, this should not be this obtuse,  Of course I don't help matters by not knowing anything related to SOAP.

I changed 
tmp['MSH']['MSH.2']="^~\\&amp;"
to
tmp['MSH']['MSH.2']="^~\\&"

This causes the XML to look like, <MSH.2>^~\&amp;amp;</MSH.2>
SOAP Sent Message,  <req:RequestHL7Message>MSH|^~\&|NJDOCCentricity

**Reply by tonygermano** [reply]
You may want to look at this https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code

I'm not sure how many of the errors in what you are copy/pasting are actually just due to formatting since you aren't using code blocks.

**Reply by LanciaH** [reply]
Great thanks, I'll take a look.

**Reply by tonygermano** [reply]
you should be able to edit your existing posts and insert the markdown to clean them up and make them more readable.

**Comment by LanciaH** [comment]
We are getting the following SOAP Error; ERROR: Error creating web service dispatch [SOAPExceptionImpl: Error during saving a multipart message]


WSDL URL: https://njiis-uat.nj.gov:443/ims/service?wsdl
SERVICE: {http://njiis.TESTCV19.gov/ims/service}NJIISIMSService
PORT / ENDPOINT: {http://njiis.TESTCV19.gov/ims/service}NJIISIMSServicePort
LOCATION URI: https://njiis-uat.nj.gov:443/ims/service
SOAP ACTION: http://njiis.nj.gov/ims/service/submitPatientImmInfo

[ATTACHMENTS]

[CONTENT]
<soap:Envelope xmlns:soap="http://www.w3.org/2003/05/soap-envelope" xmlns:req="http://njiis.TESTCV19.gov/ims/service/schema/hl7/put/request">
   <soap:Header/>
   <soap:Body>
      <req:NJIISIMSPutRequest>
         <req:RequestHeader>
            <req:FacilityID>COVID9999</req:FacilityID>
            <req:FacilityKey>COVID123$</req:FacilityKey>
         </req:RequestHeader>
         <req:PutRequest>
            
            <req:NJIISProviderID>9999</req:NJIISProviderID>
            <req:HL7MessageVersion>2.5.1</req:HL7MessageVersion>
            <req:RequestHL7Message>MSH|^~\&|NJDOCCentricity|9999|NJIIS|TESTCV19|2022042282440||VXU^V04^VXU_V04|NJDOCCentricity99992022042282440|T|2.5.1|||NE|AL|||||Z22^CDCPHINVS|PID|1||0000TESTC^^^9999^MR||TEST^TEST^^^^^L||19700606|M|||TEST WING-2 DOWN-BED 02 B^^NJ^NJ^00000^^L|||||||||||||||PD1||||||||||||Y|20211027|NK1|1|VARGAS^DAWN^^^L|OTH^SISTER IN LAW^HL70063|THOMPSON PL^^PARLIN^NJ^8859^^L|||||||ORC|RE||COVID-19^^^DOCCentricity|RXA|0|1|20211027||207^Moderna COVID-19 Vaccine^CVX|0.5|mL^^UCUM||00^New Immunization Record^NIP001|Brewer Arthur|^^^9999||||011F21A||MOD^Moderna US, Inc.^MVX|||CP|A|RXR|IM^^INTRAMUSCULAR^HL70162|RD^^RIGHT DELTOID^HL70163|OBX|1|CE|59780-7^Immunization Series Name^LN|10|850^COVID Vaccine Group^L||||||F|||||||ORC|RE||COVID-19^^^DOCCentricity|RXA|0|1|20211124||207^Moderna COVID-19 Vaccine^CVX|0.5|mL^^UCUM||00^New Immunization Record^NIP001|G KRAUS LPN|^^^9999||||011F21A||MOD^Moderna US, Inc.^MVX|||CP|A|RXR|IM^^INTRAMUSCULAR^HL70162|LA^^Left Upper Arm^HL70163|OBX|2|CE|59780-7^Immunization Series Name^LN|10|850^COVID Vaccine Group^L||||||F|||||||NTE||||</req:RequestHL7Message>
         </req:PutRequest>
      </req:NJIISIMSPutRequest>
   </soap:Body>
</soap:Envelope>

**Reply by pacmano1** [reply]
This looks like it is going to New Jersey immunization information system (IIS).

Build your tmp message normally and then enclose it is CDATA tags.    You don't need to escape anything.

![image](https://user-images.githubusercontent.com/44065187/165127522-8570bc17-94a7-42fa-86e3-667042e4e4b9.png)

**Comment by LanciaH** [comment]
Success, I think.  
In Mirth we were escaping the destination, my co worker escaped the source and the message went through.  But she thinks the data review will fail because the HL7 message not contains amp not &

See below.

<soapenv:Envelope xmlns:env="http://www.w3.org/2003/05/soap-envelope" xmlns:soapenv="http://www.w3.org/2003/05/soap-envelope">
    <env:Header/>
    <soapenv:Body xmlns:res="http://njiis.njdoh.gov/ims/service/schema/hl7/put/response">
        <res:NJIISIMSPutResponse>
            <res:ResponseHeader>
                
                <res:ServiceStatus>
                    <res:StatusCode>200</res:StatusCode>
                    <res:StatusMessage>Success</res:StatusMessage>
                </res:ServiceStatus>
            </res:ResponseHeader>
            
            <res:PutResponse>
                <res:TransactionID>PUTX9802041X20220425T121522713</res:TransactionID>
                <res:HL7MessageVersion>2.5.1</res:HL7MessageVersion>
                <res:ResponseHL7Message>MSH|^~\&amp;|NJIIS|NJDOH|

**Reply by pacmano1** [reply]
@LanciaH  Format your code, your last post should look like this:

Success, I think.
In Mirth we were escaping the destination, my co worker escaped the source and the message went through. But she thinks the data review will fail because the HL7 message not contains amp not &

See below.

```xml
<soapenv:Envelope xmlns:env="http://www.w3.org/2003/05/soap-envelope" xmlns:soapenv="[http://www.w3.org/2003/05/soap-envelope">](http://www.w3.org/2003/05/soap-envelope%22%3E)
env:Header/
<soapenv:Body xmlns:res="[http://njiis.njdoh.gov/ims/service/schema/hl7/put/response">](http://njiis.njdoh.gov/ims/service/schema/hl7/put/response%22%3E)
res:NJIISIMSPutResponse
res:ResponseHeader

res:ServiceStatus
res:StatusCode200</res:StatusCode>
res:StatusMessageSuccess</res:StatusMessage>
</res:ServiceStatus>
</res:ResponseHeader>

res:PutResponse
res:TransactionIDPUTX9802041X20220425T121522713</res:TransactionID>
res:HL7MessageVersion2.5.1</res:HL7MessageVersion>
res:ResponseHL7MessageMSH|^~&amp;|NJIIS|NJDOH|
```

**Reply by tonygermano** [reply]
Almost like that, except that's not valid xml. There's a bunch of symbols missing.

**Reply by pacmano1** [reply]
Yes, true, lol.   And I already mentioned above for immunization registries that @LanciaH can just send the HL7 in a CDATA tag.  That seems to have been ignored as well :)

**Reply by LanciaH** [reply]
Hi,
I send part messages to hide patient info,  Our next submission we are going to try the CDATA Tag, at least I think she will.

Thanks again for all the suggestions

**Comment by LanciaH** [comment]
To bring this post to a closure, my partner created a separate channel just to handle the web service part and so far it is working.  She got the idea form a youtube post.  Which leads to a different error.  

We left the & symbol double escaped and our partner accepted the data, we still don't think its right but we have to move on.
tmp['MSH']['MSH.2']= "^~\\&amp;";
![image](https://user-images.githubusercontent.com/104143486/166682581-ac130d42-7a18-4aab-b623-27cc86c71d20.png)

**Reply by tonygermano** [reply]
There are three different places escaping are happening, javascript strings, hl7 messages, and xml. In this statement `tmp['MSH']['MSH.2']= "^~\&"` it is the backslash that needs to be escaped like `tmp['MSH']['MSH.2']= "^~\\&"`, as the backslash character is the escape character for javascript strings (as well as being in the position that specifies the escape character for your HL7 message.)

The escaped string will put the value `^~\&` into MSH.2. When the ER7 HL7 message is converted to xml, the ampersand will automatically be escaped to `&amp;` as that is a requirement of xml. Any proper xml library will take an element like `<MSH.2>^~\&amp;</MSH.2>` and unescape it back to `^~\&` when extracting the text content from the element.

**Reply by LanciaH** [reply]
Thanks I'll share this with my partner.
